src/components/TodoList.tsx

```diff
@@ -0,0 +1,373 @@
+  todos, filteringMode, userId, setTodos, todosToBeEdited, setTodosToBeEdited,
+}) => {
+  const [todoTitle, setTodoTitle] = useState('');
+  const [processing, setProcessing] = useState(false);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [editingTitleTodo, setEditingTitleTodo] = useState<Todo | null>(null);
+
+  if (filteringMode !== 'all' && todos !== null) {
+    switch (filteringMode) {
+      case 'active':
+        filteredTodos = todos.filter(todo => !todo.completed);
+        break;
+      case 'completed':
+        filteredTodos = todos.filter(todo => todo.completed);
+        break;
+      default:
+    }
+  } else {
+    filteredTodos = todos;
+  }
+
+  const setError = useContext(SetErrorContext);
+
+  const handleSubmitNewTodo = (event: React.KeyboardEvent<HTMLElement>) => {
+    if (event.key === 'Enter') {
+      if (todoTitle) {
+        setProcessing(true);
+        setTempTodo({
+          title: todoTitle,
+          completed: false,
+          userId,
+          id: 0,
+        });
+        addTodo({
+          title: todoTitle,
+          completed: false,
+          userId,
+        })
+          .then((response) => {
+            setProcessing(false);
+            setTempTodo(null);
+            todos?.push(response);
+          })
+          .catch(() => setError?.(ErrorMessage.CantAdd));
+        // #TODO: get rid of the nasty ?. somehow
+
+        setTodoTitle('');
+      } else {
+        setError?.(ErrorMessage.EmptyTitle);
+        // #TODO: get rid of the nasty ?. somehow
+      }
+    }
+  };
+
+  const handleActiveToggle = (todo: Todo) => {
+    setTodosToBeEdited([todo.id]);
+    patchTodo(todo.id, { completed: !todo.completed })
+      .then(() => {
+        setTodosToBeEdited([]);
+        const todoList = [...todos];
+
+        const todoIndex = todoList.findIndex(arrTodo => arrTodo.id === todo.id);
+
+        todoList[todoIndex] = {
+          id: todo.id,
+          userId: todo.userId,
+          title: todo.title,
+          completed: !todo.completed,
+        };
+
+        setTodos(todoList);
+      })
+      .catch(() => {
+        setError?.(ErrorMessage.CantUpdate);
+        setTodosToBeEdited([]);
+      });
+  };
+
+  const handleMassActiveToggle = () => {
+    if (!todos.find(todo => !todo.completed)) {
+      setTodosToBeEdited(todos.map(todo => todo.id));
+      Promise.all(todos.map((todo) => {
+        return new Promise(
+          (resolve) => patchTodo(todo.id, { completed: !todo.completed })
+            .then(resolve),
+        )
+          .catch(() => {
+            setTodosToBeEdited([]);
+            setError?.(ErrorMessage.CantUpdate);
+          });
+      }))
+        .then(() => {
+          setTodosToBeEdited([]);
+          setTodos(todos.map(todo => {
+            return {
+              title: todo.title,
+              id: todo.id,
+              userId: todo.userId,
+              completed: !todo.completed,
+            };
+          }));
```

Comment:
also, cant you just put this code inside `finally`?

Reply:
simply changing `.then(() => {})` to `.finally(() => {})`?

technically i could do that, although im not sure what purpose would it serve